### PR TITLE
V2 : reworked toCompactBinary and fromCompactBinary functions

### LINKED ISSUE #576: V2: Errors starting web related to vtree, fails to render simple shapes
## Expected Behavior
'npm run dev' should start web and allow me to render a simple script

## Actual Behavior
Errors from missing modules in 'packages/vtree/core/modeling/primitives' 

## Steps to Reproduce the Problem
on Ubuntu 20.04
```
$ nvm use lts/carbon
Now using node v8.16.1 (npm v6.4.1)
$ git clone https://github.com/jscad/OpenJSCAD.org.git
$ cd OpenJSCAD.org; git checkout V2; npm install; npm run bootstrap; npm test
```
(all seems well)
```
$ cd packages/web
$ npm run dev
Error: Cannot find module './cylinderElliptic' from '/home/chris/lib/JSCAD-dev/OpenJSCAD.org/packages/vtree/core/modeling/primitives'
$ cd ../../packages/vtree/core/modeling/primitives
$ cp cylinder.js cylinderElliptic.js
```
(run again)
```
`Error: Cannot find module './roundedCylinder' from '/home/chris/lib/JSCAD-dev/OpenJSCAD.org/packages/vtree/core/modeling/primitives'`
(create roundedCylinder.js in packages/vtree/core/modeling/primitives based on roundedCuboid.js)
```
(run again)
```
`Error: Cannot find module './star' from '/home/chris/lib/JSCAD-dev/OpenJSCAD.org/packages/vtree/core/modeling/primitives'`
```
(copy 'roundedRectangle copy.js' to star.js in packages/vtree/core/modeling/primitives)

It runs this time!
However it fails to render any shapes with curves:
```
    return [
            torus(), 
            translate([0,5,0], cube()), 
            translate([0,10,0],cylinder()),
            translate([0,15,0], sphere())
           ]
```
Screenshot from 2020-05-11 08-22-38

so I am not sure my "fixes" are correct.

## Specifications

  - Version: V2 as of fb73090b88ca42e7534eceba73468b1944bb0806
  - Platform: Ubuntu 20.04
  - Environment: (browser, local server, node.js etc) Chromium, localhost, node v8.16.1

### LINKED ISSUE #580: objects completely broken #V2
## Expected Behavior
cylinder

## Actual Behavior
messed up triangles

## Steps to Reproduce the Problem

  1. load simple project

const jscad = require('@jscad/modeling')
const { cuboid, sphere, cylinder, circle, star } = require('@jscad/modeling').primitives
const main = () => {
  return [
    cylinder({ radius: 1 }),
  ]
}
module.exports = { main }

  2.
  3.

## Specifications

  - Version: v2
  - Platform: Windows
  - Environment: browsers: Chrome, FF

### LINKED ISSUE #582: packages/desktop `npm run dev` not working #V2
## Expected Behavior
desktop app started

## Actual Behavior
error in console
```
Error: Cannot find module 'C:\wrk\OpenJSCAD.org\packages\desktop\node_modules\.bin\electron'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:982:15)
    at Function.Module._load (internal/modules/cjs/loader.js:864:27)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:74:12)
    at internal/main/run_main_module.js:18:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
```

## Steps to Reproduce the Problem

  1. get altest openjscad V2 branch
  2. npm install
  3. cd packages/desktop
  4 . npm run dev

## Specifications

  - Version: V2
  - Platform: Windows
  - Environment: nodejs 12.16.1

### PR DESCRIPTION
**Pull request by z3dev**
These changes rework the toCompactBinary and fromCompactBinary functions.
- geom2 : convert sides, both points, to maintain direction
- geom3 : convert polygons to a list of vertices, and a list of number of vertices per polygon
- path2 : pretty much the same

There is also only one array allocation in the toCompactBinary functions, which should improve performance significantly.

And a few more tests to verify that 'empty' geometry can be converted as well.

These changes were tested in the JSCAD web application as well. All functions work fine.

Fixes #576 (rendering part)
Fixes #580 
Fixes #582 

### All Submissions:

* [X] Have you followed the guidelines in our Contributing document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [X] Does your submission pass tests?

### CODE REVIEW

**Comment by kaosat-dev** [comment]
@z3dev thanks for this, review under way, I am quite torn on a number of things tbh

**Comment by z3dev** [comment]
For geom2, the sides must be preserved as each side has order. This allows for holes. Points are not enough.

For geom3, the polygons can have 3, 4, 5, etc. points. You can turn those into triangles but you would end up with a whole lot more polygons. I opted for maintaining the complete ‘state’ of the geometry.

Colors... the color could also be converted.

**Comment by kaosat-dev** [comment]
@z3dev ok , that is along the lines of what is bothering me:
- must be exhaustion: but why are ordered points not enough ? (it adds a lot of data for the transfer from the webworker)
- geom3: humm that is bothersome, I get the logic, but again, the more data we transfer the slower it gets :( ('compact'ness of compactbinary is kinda lost)
- colors are trivial to add

Fyi I worked on a prototype a few months back that would allow to work 100% in web workers (no data ever transfered back to main thead), so it would remove the need for all these conversions, unfortunatly , it depends on some browser features that are not 100% present

**Comment by z3dev** [comment]
>     * must be exhaustion: but why are ordered points not enough ? (it adds a lot of data for the transfer from the webworker)

geom2 with holes. that's the main reason to keep the sides.

>     * geom3: humm that is bothersome, I get the logic, but again, the more data we transfer the slower it gets :( ('compact'ness of compactbinary is kinda lost)

Understand exactly. I didn't really want to make these changes for geom3. It seems like a waste of time going from one array to another array. There are no 'planes' so that saves data transfers.

> Fyi I worked on a prototype a few months back that would allow to work 100% in web workers (no data ever transfered back to main thead), so it would remove the need for all these conversions, unfortunatly , it depends on some browser features that are not 100% present

Yeah. It certianly would be nice if all browsers maintained the same level of JS standards.

File: packages/modeling/src/geometry/geom2/fromCompactBinary.js [hunk 1]
```diff
@@ -1,23 +1,27 @@
...
-    pts.push([dat[i], dat[i + 1]])
+const fromCompactBinary = data => {
+  if (data[0] !== 0) throw new Error('invalid compact binary data')
+
+  const created = create()
+
+  created.transforms = mat4.clone(data.slice(1, 17))
+
+  for (let i = 17; i < data.length; i += 4) {
+    const point0 = vec2.fromValues(data[i + 0], data[i + 1])
```
[the remark this answers is not in the document]
Comment (z3dev) [reply packages/modeling/src/geometry/geom2/fromCompactBinary.js:20]: Good point.

File: packages/modeling/src/geometry/geom2/fromCompactBinary.js [hunk 2]
```diff
@@ -1,23 +1,27 @@
-// FIXME: how about custom properties or fields ?
```
[the remark this answers is not in the document]
Comment (z3dev) [reply packages/modeling/src/geometry/geom2/fromCompactBinary.js:1]: I added TODOs for those custom properties.
Done.

File: packages/modeling/src/geometry/geom2/fromCompactBinary.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (z3dev) [reply packages/modeling/src/geometry/geom2/fromCompactBinary.js:20]: OK. The use of vec2/vec3 ensures that the array is of the correct type, i.e. Float32Array.

By switching to [data[i + 0], data[i + 1]] then the type will be that of Number. There's still an allocation of an array.

It's probably better to use vec2/vec3.
